libraries/chain/webassembly/eos-vm-oc/executor.cpp

```diff
@@ -0,0 +1,251 @@
+#include <eosio/chain/webassembly/eos-vm-oc/executor.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/code_cache.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/memory.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic.hpp>
+#include <eosio/chain/wasm_eosio_constraints.hpp>
+#include <eosio/chain/apply_context.hpp>
+#include <eosio/chain/transaction_context.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/types.hpp>
+
+#include <fc/scoped_exit.hpp>
+
+#include <boost/hana/equal.hpp>
+
+#include <mutex>
+
+#include <asm/prctl.h>
+#include <sys/prctl.h>
+#include <sys/syscall.h>
+
+extern "C" int arch_prctl(int code, unsigned long* addr);
+
+namespace eosio { namespace chain { namespace eosvmoc {
+
+static constexpr auto signal_sentinel = 0x4D56534F45534559ul;
+
+static std::mutex inited_signal_mutex;
+static bool inited_signal;
+
+static void(*chained_handler)(int,siginfo_t*,void*);
+[[noreturn]] static void segv_handler(int sig, siginfo_t* info, void* ctx)  {
+   control_block* cb_in_main_segment;
+
+   //a 0 GS value is an indicator an executor hasn't been active on this thread recently
+   uint64_t current_gs;
+   syscall(SYS_arch_prctl, ARCH_GET_GS, &current_gs);
```

Comment:
What is the rationale for using syscall over the glibc wrapper?

Reply:
That the docs don't allow `arch_prctl` & `mprotect` from a signal handling context. Yeah the docs don't allow `syscall` either but it's even less likely to do be touching some state.

Reply:
Setting errno is an issue, so I suppose I can live with this in the signal handler.  You're using syscall in other places, too, though.